docs/_includes/Types.md

```diff
@@ -974,6 +975,99 @@ if K in cache {  // check if temperature is in domain of cache
 }
 ```
 
+## 10.5. Iterating over collections
+
+Collections are very commonly used in programming and one frequently
+needs to iterate over the elements of a collection. Dafny does not have
+built-in iterator methods, but the idioms by which to do so are straightforward.
+The subsections below give some introductory examples; more
+detail can be found in this [power user note](http://leino.science/papers/krml275.html).
+
+TODO: Add examples of using a iterator class
+TODO: Should a foreach statment be added to Dafny
+
+### 10.5.1. Sequences and arrays
+
+Sequences and arrays are indexable and have a length. So the idiom to
+iterate over the contents is well-known. For an array:
+```dafny
+  var i: int := 0;
+  var sum: int := 0;
+  while i < s.Length
+    decreases s.Length - i
+  {
+    sum := sum + s[i];
+    i := i + 1;
+  }
+}
+```
+For a sequence, the only difference is the length operator:
+```dafny
+  var i: int := 0;
+  var sum: int := 0;
+  while i < |s|
+    decreases |s| - i
+  {
+    sum := sum + s[i];
+    i := i + 1;
+  }
+```
+
+The Forall statement ([Section 22.8](#sec-forall-statement)) can also be used
+with arrays where parallel assigment is needed:
+```dafny
+  var rev := new int[s.Length];
+  forall i | 0 <= i < s.Length {
+    rev[i] := s[s.Length-i-1];
+  }
+```
+
+### 10.5.2. Sets
+There is no intrinsic order to the elements of a set. Nevertheless, we can
+extract an arbitrary element of a non-empty set, performing an iteration
+as follows:
+```dafny
+// s is a set<int>
+  var ss := s;
+  while ss != {}
+    decreases |ss|
+  {
+    var i: int :| i in ss;
+    ss := ss - {i};
+    print i, "\n";
+  }
+```
+
+A comparable loop can be written for an `iset`, though it may not terminate:
+```dafny
+  var ss: iset<int> := s;
+  while (ss != iset{})
+    decreases *
+  {
+    var i: int :| i in ss;
```

Comment:
Actually, this is not allowed, either, for the same reason that you can't iterate over an `imap`.

Reply:
The complaint from Dafny is obscure then: Error: this assign-such-that statement is too advanced for the current compiler; Dafny's heuristics cannot find any bound for variable 'I'

I'll remove the example.